### ISSUE TITLE
`torch.compile` silently produces wrong results for `index_add`/`index_copy`/`index_fill`/`index_put` on expanded (stride-0) tensors

### ISSUE THREAD

**Issue by wuyii8941** [pytorch/pytorch#183986]
### 🐛 Describe the bug

When a tensor is created via `.expand()` (introducing stride-0 dimensions), `index_add`, `index_copy`, `index_fill`, and `index_put` under `torch.compile(backend='inductor')` produce **silently wrong forward results**. The compiled code operates directly on the aliased storage instead of materializing the expansion, so writes intended for independent rows collide through the shared underlying memory.

Eager mode and `aot_eager` are correct. Only the Inductor backend is affected.

### Reproducer

```python
import torch

def fn(src, idx):
    x = torch.zeros(1, 8, device="cuda").expand(4, -1)  # stride-0 in dim 0
    return x.index_add(0, idx, src)

src = torch.ones(3, 8, device="cuda")
idx = torch.tensor([0, 1, 0], device="cuda")

eager = fn(src, idx)
compiled = torch.compile(fn, backend="inductor")(src, idx)

print("Eager row 0:", eager[0].tolist())      # [2, 2, 2, 2, 2, 2, 2, 2]
print("Compiled row 0:", compiled[0].tolist()) # [3, 3, 3, 3, 3, 3, 3, 3]  ← WRONG
print("Eager row 2:", eager[2].tolist())       # [0, 0, 0, 0, 0, 0, 0, 0]
print("Compiled row 2:", compiled[2].tolist()) # [3, 3, 3, 3, 3, 3, 3, 3]  ← WRONG
```

**Expected:** Row 0 = `[2, ...]` (idx 0 hit twice), Row 1 = `[1, ...]`, Rows 2–3 = `[0, ...]`
**Actual (compiled):** All rows = `[3, ...]` — the total sum leaks to every row via aliased storage.

### All affected ops

```python
import torch

def test_op(name, op_fn):
    torch._dynamo.reset()
    eager = op_fn()
    compiled = torch.compile(op_fn, backend="inductor")()
    diff = (eager - compiled).abs().max().item()
    print(f"  {name}: max_diff={diff:.4f}  {'BUG' if diff > 0.01 else 'OK'}")

idx = torch.tensor([0, 1, 0], device="cuda")
src = torch.ones(3, 8, device="cuda")

test_op("index_add", lambda: torch.zeros(1, 8, device="cuda").expand(4, -1).index_add(0, idx, src))
test_op("index_copy", lambda: torch.zeros(1, 8, device="cuda").expand(4, -1).index_copy(0, idx[:2], src[:2]))
test_op("index_fill", lambda: torch.zeros(1, 8, device="cuda").expand(4, -1).index_fill(0, idx[:2], 1.0))
test_op("index_put(acc=True)", lambda: torch.zeros(1, 8, device="cuda").expand(4, -1).index_put((idx,), src, accumulate=True))
test_op("index_put(acc=False)", lambda: torch.zeros(1, 8, device="cuda").expand(4, -1).index_put((idx[:2],), src[:2], accumulate=False))
# All print BUG
```

`scatter`, `scatter_add`, `scatter_reduce`, `select_scatter`, `slice_scatter` are NOT affected — they go through a different codegen path that correctly handles stride-0 inputs.

### Controls

| Condition | Result |
|-----------|--------|
| Non-expanded tensor (`torch.zeros(4, 8)`) | Correct (diff=0) |
| `.clone()` after `.expand()` | Correct (diff=0) |
| `aot_eager` backend | Correct (diff=0) |
| All shapes tested (4×8 to 128×256) | All wrong |
| All dtypes (float16, float32, float64) | All wrong |
| Expand on dim 0 or dim 1 | Both wrong |
| Backward (src.grad) | Correct — only forward is affected |

### Root cause

Inductor's lowering for the `index_*` family does not detect or materialize expanded (stride-0) tensors before performing indexed writes. Because all expanded rows alias the same storage, writes to different logical rows collide.

### Workaround

Insert `.clone()` or `.contiguous()` after `.expand()`:
```python
x = torch.zeros(1, 8, device="cuda").expand(4, -1).clone()  # materialize
```

### Relationship to existing issues

This is the same bug class as #180164 / #182012 (stride-0 tensor mishandling in Inductor), but:
- #180164 (`slice_scatter`/`select_scatter` backward): **fixed**
- #182012 (`diagonal_scatter` backward): **open, fix PR ready**
- This bug: `index_add`/`index_copy`/`index_fill`/`index_put` **forward** — NOT fixed

The `index_*` ops are a distinct code path from the `*_scatter` ops, so existing fixes do not cover them.

### Versions

- PyTorch: 2.13.0.dev20260513+cu126
- GPU: Tesla T4 (SM 7.5), also reproducible on CPU
- OS: Linux

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by puneetdixit200** [comment]
I’d like to work on this correctness issue. I can add an Inductor regression test for expanded stride-0 tensors with index_add/index_copy/index_fill/index_put, then trace the lowering that loses eager semantics. Please assign this to me.

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260514+cu130`) on NVIDIA H100 and confirmed it reproduces (3/3 runs failed).

All 3 runs failed with consistent numerical mismatch on index_add over expanded (stride-0) tensor. Eager row 0=[2,...], compiled row 0=[3,...]; max_diff=3.0. Diagnostic sweep shows all 5 affected ops from the issue (index_add, index_copy, index_fill, index_put accumulate=True/False) produce wrong results, while eager is correct. emulate_precision_casts=1 mitigation did not fix it.

<details>
<summary>Error output</summary>

```
Device: cuda
Torch: 2.13.0.dev20260514+cu130
GPU: NVIDIA H100

All affected ops:
  index_add: max_diff=3.0000  BUG
  index_copy: max_diff=1.0000  BUG
  index_fill: max_diff=1.0000  BUG
  index_put(acc=True): max_diff=3.0000  BUG
  index_put(acc=False): max_diff=1.0000  BUG

Primary repro (index_add):
  Eager row 0:    [2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0]
  Compiled row 0: [3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0]
  Eager row 2:    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
  Compiled row 2: [3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0]
  max_diff: 3.0

ERROR conda.cli.main_run:execute(49): `conda run python repro.py` failed. (See above for error)
```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import torch


def fn(src, idx):
    x = torch.zeros(1, 8, device="cuda").expand(4, -1)
    return x.index_add(0, idx, src)


def test_fn() -> bool:
    """Return True if NO bug (eager == compiled). False if bug present."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")

    device = "cuda" if torch.cuda.is_available() else "cpu"
    src = torch.ones(3, 8, device=device)
    idx = torch.tensor([0, 1, 0], device=device)

    # eager output
    eager_out = fn(src, idx)

    # compiled output
    compiled_fn = torch.compile(fn, backend=backend)
    compiled_out = compiled_fn(src, idx)

    return torch.allclose(eager_out, compiled_out, atol=1e-4)


if __name__ == "__main__":
    device = "cuda" if torch.cuda.is_available() else "cpu"
    print(f"Device: {device}")
    print(f"Torch: {torch.__version__}")
    if torch.cuda.is_available():
        print(f"GPU: {torch.cuda.get_device_name(0)}")

    # Diagnostic exploration of the bug across all reported affected ops
    def diag_test_op(name, op_fn):
        torch._dynamo.reset()
        eager = op_fn()
        compiled = torch.compile(op_fn, backend="inductor")()
        diff = (eager - compiled).abs().max().item()
        status = "BUG" if diff > 0.01 else "OK"
        print(f"  {name}: max_diff={diff:.4f}  {status}")

    idx = torch.tensor([0, 1, 0], device=device)
    src = torch.ones(3, 8, device=device)
    print("\nAll affected ops:")
    diag_test_op("index_add", lambda: torch.zeros(1, 8, device=device).expand(4, -1).index_add(0, idx, src))
    diag_test_op("index_copy", lambda: torch.zeros(1, 8, device=device).expand(4, -1).index_copy(0, idx[:2], src[:2]))
    diag_test_op("index_fill", lambda: torch.zeros(1, 8, device=device).expand(4, -1).index_fill(0, idx[:2], 1.0))
    diag_test_op("index_put(acc=True)", lambda: torch.zeros(1, 8, device=device).expand(4, -1).index_put((idx,), src, accumulate=True))
    diag_test_op("index_put(acc=False)", lambda: torch.zeros(1, 8, device=device).expand(4, -1).index_put((idx[:2],), src[:2], accumulate=False))

    print("\nPrimary repro (index_add):")
    torch._dynamo.reset()
    src = torch.ones(3, 8, device=device)
    idx = torch.tensor([0, 1, 0], device=device)
    eager = (torch.zeros(1, 8, device=device).expand(4, -1)).index_add(0, idx, src)
    compiled = torch.compile(lambda s, i: torch.zeros(1, 8, device=device).expand(4, -1).index_add(0, i, s), backend="inductor")(src, idx)
    print(f"  Eager row 0:    {eager[0].tolist()}")
    print(f"  Compiled row 0: {compiled[0].tolist()}")
    print(f"  Eager row 2:    {eager[2].tolist()}")
    print(f"  Compiled row 2: {compiled[2].tolist()}")
    print(f"  max_diff: {(eager - compiled).abs().max().item()}")

    try:
        ok = test_fn()
    except Exception:
        import traceback
        traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260514+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.13.0.dev20260514+cu130
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260515+cu130
[pip3] torchdata==0.11.0
[pip3] torchmetrics==1.0.3
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260515+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
